server/src/main/java/com/linecorp/centraldogma/server/internal/storage/project/DefaultProject.java

```diff
@@ -196,11 +199,63 @@ private void initializeMetadata(long creationTimeMillis, Author author) {
                                                                  ImmutableMap.of(member.id(), member),
                                                                  ImmutableMap.of(),
                                                                  userAndTimestamp, null);
+            final CommitResult result =
+                    dogmaRepo.commit(headRev, creationTimeMillis, Author.SYSTEM,
+                                     "Initialize metadata", "",
+                                     Markup.PLAINTEXT,
+                                     Change.ofJsonUpsert(METADATA_JSON, Jackson.valueToTree(metadata)))
+                             .join();
+            lastMetadataRevision = result.revision();
+            this.projectMetadata = metadata;
+        }
+    }
 
-            dogmaRepo.commit(headRev, creationTimeMillis, Author.SYSTEM,
-                             "Initialize metadata", "", Markup.PLAINTEXT,
-                             Change.ofJsonUpsert(METADATA_JSON, Jackson.valueToTree(metadata))).join();
+    @Nullable
+    private ProjectMetadata initialMetadata()
+            throws ExecutionException, InterruptedException, JsonProcessingException {
+        if (name.equals(INTERNAL_PROJECT_DOGMA)) {
+            return null;
+        }
+        final Entry<JsonNode> metadata = repos.get(REPO_DOGMA).get(Revision.HEAD, Query.ofJson(METADATA_JSON))
+                                              .get();
+        final ProjectMetadata projectMetadata = Jackson.treeToValue(metadata.content(),
+                                                                    ProjectMetadata.class);
+        lastMetadataRevision = metadata.revision();
+        this.projectMetadata = projectMetadata;
+        return projectMetadata;
+    }
+
+    /**
+     * Listens to new changes for "metadata.json" and updates the information to {@link #lastMetadataRevision}
+     * and {@link #projectMetadata}.
+     */
+    private void attachMetadataListener() {
+        if (name.equals(INTERNAL_PROJECT_DOGMA)) {
+            return;
         }
+
+        final Repository dogmaRepo = repos.get(REPO_DOGMA);
+        dogmaRepo.addListener(RepositoryListener.of(Query.ofJson(METADATA_JSON), entry -> {
+            if (entry == null) {
+                logger.warn("{} file is missing in {}/{}", METADATA_JSON, name, REPO_DOGMA);
+                return;
+            }
+
+            final Revision lastRevision = entry.revision();
+            if (lastRevision.compareTo(lastMetadataRevision) <= 0) {
```

Comment:
nit: `assert lastMetadataRevision != null;`
We might want to assign it to a local variable not to access the field twice.